### PR TITLE
support for BaseTool2 in Ultimate tool-info-modules

### PR DESCRIPTION
**Pull request by danieldietsch**

### CODE REVIEW

**Comment by danieldietsch** [comment]
@skanav  this should probably take care of the ultimate-based tools (see also #646)

**Comment by danieldietsch** [comment]
Ah I just realized that I also need to take care of the data model, not just update to ``BaseTool2``

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:80]: Note that tool-info modules based on `BaseTool2` should not look for the tool in the current directory automatically anymore. `ToolLocator` will search in the current directory if and only if this is desired according to the BenchExec configuration.

And AFAIS this should mean that the whole method can and should be simplified to a simple call to `tool_locator.find_executable` as in most other tool-info modules.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:208]: I did not notice this before, but storing information about a task in `self` is broken, because there is no guarantee that `cmdline` and `determine_result` are called in any particular order (cf. docs of `BaseTool2`). So please remove this field. With the new API this should finally be possible because `determine_result` has access to the run command line.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:452]: `None` is in general a better return value. Using a string will lead to table columns with mixed numbers and strings, so there won't be any statistics etc. With `None` the respective cells will be correctly interpreted as "missing value".

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:68]: No reason to cache this, actually.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:216]: Note that the new API explicitly guarantees that information about the tool version may be stored in a field of `self` and used by methods like `cmdline`, so if you wish, you could simplify this.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:358]: Just noticed this: Looking for strings in Python is most easily expressed as `if needle in haystack:`

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:20]: For the cases where this exception is currently used, `ToolNotFoundException` seems a good replacement.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:69]: I am wondering whether this is as intended. `os.walk` walks recursively through directories, so the check will succeed if there are `Ultimate` and `plugins` directory anywhere below the executable's directory. The error message seems to indicate that all three are expected to be in the same directory instead?

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:190]: `task.options` is not a list of command-line parameters, but an arbitrary data structure (as defined in the task-definition file in the `options` key). So this needs to be converted to command-line parameters first. Cf. #646 and the `sv_benchmarks_util` that is added there for code that is useful for most cases. Maybe we should delay this PR until #646 is merged and then rebase on top of it?

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:212]: Simplification:
```suggestion
        if task.property_file:
```

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:219]: `task.property_file` is guaranteed to be `None` here, right?

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:234]: I noticed this code now and while the decision of how to do this is up to you, I wanted to mention the following (from my experience as a CPAchecker developer, not as BenchExec maintainer):

The memory limit from BenchExec (which is in `mem_bytes`) is a hard limit, reaching it will let the process be killed. Using it as the max heap size of a JVM will mean that the JVM might try to use that much memory for heap plus all the other memory (native allocations, stack, etc.), so it might well exceed the limit. Due to the GC, the JVM could even try to allocate that much memory even if it could cope with less memory (at the cost of more GC runs). So in CPAchecker, for example, we always set the max heap to some smaller value. How much smaller unfortunately depends on a lot of things, including how much memory is needed by non-JVM allocations (e.g., from a native SMT solver).

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:248]: Seems like `UnsupportedFeatureException` would be a good fit here and in similar locations?

Of course you could in principle also remove these checks and just let Ultimate print an error message when such a command line is attempted to be executed. It would be a little less convenient for the user, but reduce the maintenance effort of the tool-info module for you. So I leave this up to you.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:272]: `task.input_files` will always be non-empty when it is returned, otherwise an exception is thrown. If you want to support both empty and non-empty list of input files, you would need to use
```suggestion
        if task.input_files_or_empty:
```

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:273]:
```suggestion
            cmdline += ["-i", *task.input_files]
```

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:280]: Same as above.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:292]: This is enough:
```suggestion
        cmdline += task.input_files
```

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:314]:
```suggestion
        if any("--spec" in arg or ".prp" in arg for arg in run.commandline):
```

But do you actually want to the check whether there is one argument that contains `--spec`, or whether there is one argument that is equal to `--spec`?
If the latter:
```suggestion
        if any(arg == "--spec" or ".prp" in arg for arg in run.commandline):
```

**Review by PhilippWendler — CHANGES_REQUESTED** [review]
Looks mostly good, but the `task.options` need to be handled differently and I have a few other non-critical questions and suggestions.

File: benchexec/tools/ultimate.py
Comment (danieldietsch) [reply benchexec/tools/ultimate.py:69]: But there is an unconditional ``break`` after the ``if``

File: benchexec/tools/ultimate.py
Comment (danieldietsch) [reply benchexec/tools/ultimate.py:190]: That sounds good! Lets do that.

File: benchexec/tools/ultimate.py
Comment (danieldietsch) [reply benchexec/tools/ultimate.py:234]: Yes, we had the same policy with Ultimate for a long time. Nevertheless, it seems that allowing the VM in prinicpal to go up to this bound reduces OOMs, at least on SVCOMP instances.

File: benchexec/tools/ultimate.py
Comment (danieldietsch) [reply benchexec/tools/ultimate.py:248]: Yes, but I cannot change older Ultimate versions. So to be compatible with them we have a complex tool info module ;)

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [reply benchexec/tools/ultimate.py:69]: Ok, right. Hard to spot :-)

How about
```
directory = os.path.dirname(exe)
if os.path.isfile(os.path.join(directory, "Ultimate")) and os.path.isdir(os.path.join(directory, "plugins")):
```
instead of the loop?
I think this would be better understandable.

But I am leaving this up to you.

Comment (danieldietsch) [reply benchexec/tools/ultimate.py:69]: Ah, I leave it as is; it might be that we need to check for even more files in the near future.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [reply benchexec/tools/ultimate.py:190]: #646 is merged now.

**Comment by danieldietsch** [comment]
@PhilippWendler I get the following exception on my Python 3.7 when I run ``python3 -m benchexec.test_tool_info ultimateautomizer --no-container --tool-directory /storage/repos/ultimate/releaseScripts/default/UAutomizer-linux``

My python is 3.7.8. 

```
WARNING: Tool-info module does not support SV-Benchmarks task: “expected str, bytes or os.PathLike object, not list”
Traceback (most recent call last):
  File "/storage/repos/benchexec/benchexec/test_tool_info.py", line 108, in log_if_unsupported
    yield  # call code block to be executed
  File "/storage/repos/benchexec/benchexec/test_tool_info.py", line 272, in print_standard_task_cmdlines
    CURRENT_BASETOOL.ResourceLimits(cputime=900, cputime_hard=1000),
  File "/storage/repos/benchexec/benchexec/model.py", line 219, in cmdline_for_run
    args = [os.path.expandvars(arg) for arg in args]
  File "/storage/repos/benchexec/benchexec/model.py", line 219, in <listcomp>
    args = [os.path.expandvars(arg) for arg in args]
  File "/usr/lib/python-exec/python3.7/../../../lib/python3.7/posixpath.py", line 288, in expandvars
    path = os.fspath(path)
TypeError: expected str, bytes or os.PathLike object, not list
```

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:198]: Either change this line to `combined_options += ["--architecture", data_model_param]"` and make the values of the above dict to simple strings, or change this line to `combined_options += data_model_param` (though in the later case, `data_model_param` should really be named `data_model_params`.

This is the cause of the crash due to a nested list being contained in the result value.

File: benchexec/tools/predatorhp.py
Comment (PhilippWendler) [inline benchexec/tools/predatorhp.py:34]: Please don't reformat other files.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:197]: This check would mean if a user manually wants to override the architecture of a task by adding `--architecture something` to the command line, the tool-info module will still add `--architecture` as well, leading to duplicate parameters.

I would suggest to simplify this and simply check wether `--architecture` exists in `options`. This is also how the other tool-info modules are currently implemented.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:321]:
```suggestion
        if any(arg for arg in run.commandline if "--spec" == arg or ".prp" in arg):
```
The tool `flake8` that we use for checking best practices of Python code suggests this, and we use it in CI so we want to apply all of its suggestions when they make sense.

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:307]: `.format()` does not support `%s`, so this needs to changed:
```suggestion
        ), "cmdline contains empty or None argument when using {} mode: {}".format(
```

**Review by PhilippWendler — CHANGES_REQUESTED** [review]

File: benchexec/tools/predatorhp.py
Comment (danieldietsch) [reply benchexec/tools/predatorhp.py:34]: Oh, sorry, was not my intention ...

File: benchexec/tools/ultimate.py
Comment (danieldietsch) [reply benchexec/tools/ultimate.py:321]: Do you have a flake8 config somewhere so that I can integrate it into PyCharm? I ran flake8 3.8.4, and it did not complain about that.

Comment (PhilippWendler) [reply benchexec/tools/ultimate.py]: Sure, our config is checked in already (`.flake8`). It just does not run on Travis CI due to the inflexibility of Travis' configuration, and GitLab CI unfortunately does not work on GitHub PRs yet. This is why I mention it manually.

If you want to check it for yourself, this is of course nice, and you are probably just missing the package `flake8-awesome` with additional plugins. In our `.gitlab-ci.yml` for job `flake8` it is visible that there is nothing necessary besides installing that package and running `flake8`. (Of course there is no expectation for PR authors to look there, I am just mentioning this as a possible reference.)

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [reply benchexec/tools/ultimate.py]: A warning like you added is of course even nicer for the user!

File: benchexec/tools/ultimate.py
Comment (PhilippWendler) [inline benchexec/tools/ultimate.py:193]:
```suggestion
            task, {ILP32: "32bit", LP64: "64bit"}
```

The version of black used by us would otherwise reformat this. I would actually have committed this myself but could not push to the PR branch, so please apply this suggestion.

**Review by PhilippWendler — APPROVED** [review]
Thank you very much for this improvement!

There is just a small technicality that needs to be done before the merge.

**Pull request merged**
